libr/util/event.c

```diff
@@ -0,0 +1,37 @@
+/* radare2 - MIT - Copyright 2018 - pancake */
+
+#include <r_util.h>
+
+R_API REvent *r_event_new(void *user) {
+	REvent *ev = R_NEW0 (REvent);
+	if (ev) {
+		ev->callbacks = r_list_newf (NULL);
```

Comment:
more than a list, I think we need an hashtable (you can use `ht_new`). If a listener register itself for the event R_EVENT_META_SET, there's no reason why it should be called for other kind of events too.

Reply:
what if you want to hook all events? my idea was to filter the event type in the hook itself. because i dont think we'll need to add more than 10 event hooks, unless we abuse the event system for other things like rbinpull or so

Reply:
What about a bitmask with different event types? This obviously limits us to 32/64 types, but maybe some can be put together to an event "scope".

Reply:
we could just use a ALL type during r_event_hook and just register the listenere in all type of events.

Reply:
Also, I think instead of a list I would intuitively use RVector. I wouldn't expect callbacks to be added or removed randomly all the time, but you will iterate the list very often, so RVector will probably be the better choice.

Reply:
i think we will have more than 64 types of events at some point.. unless we do sub-types

Reply:
Why not just the hashtable? It's really simple to create/iterate/whatever. It doesn't limit the number of events we can have, it's fast, it's fast to get the listeners from a given type...

Reply:
My point not about the hashtable, what I meant was if the question is RList vs. RVector, then RVector is probably the better choice here.

How would you do the hashtable exactly? REventType -> List of Callbacks? Then how do we handle the case that you want to subscribe to multiple or all events?

Reply:
```C
int add_callback(void *user, const char *key, const void *value) {
    RVector *v = (RVector *)value;
    REventCallback *cb = (REventCallback *)user;
    r_vector_push (v, cb);
    return 1;
}

R_API void r_event_hook(REvent *ev, REventType type, REventCallback cb) {
    if (type == R_EVENT_ALL) {
        ht_foreach (ev->callbacks, add_callback, cb);
    } else {
        RVector *v = ht_find (ev->callbacks, type, NULL);
        r_vector_push (v, cb);
    }
}
```

Reply:
Yep, lgtm too. This should make it quite efficient to call the callbacks.

Reply:
agree on vector